#!/bin/bash
#PBS -N bwa-mem2_align
#PBS -l walltime=24:00:00
#PBS -l select=1:ncpus=32:mem=128gb
#PBS -J 0-5

#change work directory to submission directory
cd $PBS_O_WORKDIR

#load and activate conda work environment
eval "$(~/miniforge3/bin/conda shell.bash hook)"
conda activate hpcdemo

#SRR names via csv file
csv_path="$HOME/atac_seq/ren_et_al_2021/ren_sra_atac.csv"
SRR=($(tail -n +2 "$csv_path" | cut -d ',' -f 1 ))


#paths to sequencing reads
read1="$EPHEMERAL/atac_seq/ren_et_al_2021/fastq_trimmed_3/${SRR[$PBS_ARRAY_INDEX]}_1_trimmed.fastq"
read2="$EPHEMERAL/atac_seq/ren_et_al_2021/fastq_trimmed_3/${SRR[$PBS_ARRAY_INDEX]}_2_trimmed.fastq"


#path to blacklist .bed file
blacklist="$HOME/BLACKLIST/hg38-blacklist.v2.bed"

#ref fa file
ref_fa="$HOME/GENCODE_GENOME_INDEX_100bp/GRCh38.primary_assembly.genome.fa"

#output of bam files
bam="$EPHEMERAL/atac_seq/ren_et_al_2021/bam/"
mkdir -p $bam

trimmed_bam="$EPHEMERAL/atac_seq/ren_et_al_2021/trimmed_bam/"
mkdir -p $trimmed_bam

sorted_trimmed_bam="$EPHEMERAL/atac_seq/ren_et_al_2021/sorted_trimmed_bam/"
mkdir -p $sorted_trimmed_bam

# chr list
chr="$HOME/scripts/chip-seq/g_coord.txt"

#### PROCESS CODE #########

# FIRST: bwa-mem2 alignment and pipe into a bam file
#bwa-mem2 mem -t 32 "$HOME/BWA-MEM2-INDEX/reference_index" $read1 $read2 | samtools sort -o "$bam/${SRR[$PBS_ARRAY_INDEX]}_s.bam" -

# SECOND: remove blacklisted regions
bedtools subtract -a "$bam/${SRR[$PBS_ARRAY_INDEX]}_s.bam" -b "$blacklist" > "$trimmed_bam/${SRR[$PBS_ARRAY_INDEX]}_t.bam"

# THIRD: removal of 0 coordinates (unmapped reads/improper alignment)
samtools view -@ 16 -q 1 -t "$ref_fa" -b "$trimmed_bam/${SRR[$PBS_ARRAY_INDEX]}_t.bam" | samtools sort -@ 16 -o "$sorted_trimmed_bam/${SRR[$PBS_ARRAY_INDEX]}_ts.bam" -

# FOURTH: index using samtools
samtools index "$sorted_trimmed_bam/${SRR[$PBS_ARRAY_INDEX]}_ts.bam"

conda deactivate
