#!/bin/bash
#PBS -N STAR_index
#PBS -l walltime=08:00:00
#PBS -l select=1:ncpus=32:mem=128gb
#PBS -J 0-15

#change work directory to submission directory
cd $PBS_O_WORKDIR

#loading the Conda environment:
eval "$(~/miniforge3/bin/conda shell.bash hook)"

#activate conda environment
source activate hpcdemo

########################## CREATING GENOME INDEX ######################################
STAR --runThreadN 8 --runMode genomeGenerate \
--genomeDir "$HOME/GENCODE_GENOME_INDEX" \
--genomeFastaFiles "$HOME/GENCODE_GENOME_INDEX/GRCh38.primary_assembly.genome.fa" \
--sjdbGTFfile "$HOME/GENCODE_GENOME_INDEX/basic_gene_annotation_PRI.gtf" \
--sjdbOverhang 99


################################# ALIGNMENT ###########################################
#SRR names via csv file
csv_path="$HOME/rna_seq/mcd_et_al_2017/SraRunTable_macdonald2017_rna.csv"
SRR=($(tail -n +2 "$csv_path" | cut -d ',' -f 1 ))

#file paths to sequencing reads
file_name_R1="${SRR[$PBS_ARRAY_INDEX]}_1.fastq"
file_name_R2="${SRR[$PBS_ARRAY_INDEX]}_2.fastq"

#set the paths for GTF file and outpaths
inpath=$EPHEMERAL/rna_seq/mcd_et_al_2017/mcd_fastq
outpath=$EPHEMERAL/rna_seq/mcd_et_al_2017/mcd_bam
ano_gtf_path=$HOME/GENCODE_GENOME_INDEX_100bp/basic_gene_annotation_PRI.gtf

#STAR code - note, this should be run as one whole command, not one command per line!
STAR --genomeDir "$HOME/GENCODE_GENOME_INDEX_100bp" --runThreadN 32 --limitBAMsortRAM >= 0 --sjdbGTFfile $ano_gtf_path --readFilesIn "$inpath/$file_name_R1" "$inpath/$file_name_R2" --quantMode GeneCounts --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded --outFileNamePrefix "${outpath}/${SRR[$PBS_ARRAY_INDEX]}"  --outSAMtype BAM SortedByCoordinate --outSAMunmapped Within --outSAMattributes Standard \

conda deactivate
